File: 2026-01/src/day_28/boj_9328/Main.java
package day_28.boj_9328;

import java.io.*;
import java.util.*;

public class Main {

    static int[] dr = {-1, 0, 1, 0};
    static int[] dc = {0, 1, 0, -1};
    static int n, m;
    static char[][] grid;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        int t = Integer.parseInt(br.readLine());
        while (t-- > 0) {
            st = new StringTokenizer(br.readLine());
            int h = Integer.parseInt(st.nextToken());
            int w = Integer.parseInt(st.nextToken());

            n = 1 + h + 1;
            m = 1 + w + 1;

            grid = new char[n][m];
            for (int i = 0; i < n; i++) {
                grid[i][0] = grid[i][m - 1] = '.';
            }
            for (int j = 0; j < m; j++) {
                grid[0][j] = grid[n - 1][j] = '.';
            }

            for (int i = 1; i <= h; i++) {
                String input = br.readLine();
                for (int j = 1; j <= w; j++) {
                    grid[i][j] = input.charAt(j - 1);
                }
            }

            boolean[] hasKey = new boolean[26];

            String s = br.readLine();
            if (!s.equals("0")) {
                for (char c : s.toCharArray()) {
                    hasKey[c - 'a'] = true;
                }
            }

            sb.append(bfs(hasKey)).append("\n");
        }

        System.out.println(sb);
    }

    static boolean isKey(char c) {
        return 'a' <= c && c <= 'z';
    }

    static boolean isDoor(char c) {
        return 'A' <= c && c <= 'Z';
    }

    static int bfs(boolean[] hasKey) {
        Queue<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{0, 0});

        Queue<int[]>[] doors = new ArrayDeque[26];
        for (int i = 0; i < 26; i++) {
            doors[i] = new ArrayDeque<>();
        }

        boolean[][] vis = new boolean[n][m];
        vis[0][0] = true;

        int cnt = 0;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            if (grid[cur[0]][cur[1]] == '$') cnt++;

            for (int d = 0; d < 4; d++) {
                int nr = cur[0] + dr[d];
                int nc = cur[1] + dc[d];

                if (nr < 0 || nr >= n || nc < 0 || nc >= m) continue;
                if (grid[nr][nc] == '*' || vis[nr][nc]) continue;

                if (isDoor(grid[nr][nc])) {
                    if (!hasKey[grid[nr][nc] - 'A']) {
                        doors[grid[nr][nc] - 'A'].offer(new int[]{nr, nc});
                        vis[nr][nc] = true;
                        continue;
                    }
                }

                if (isKey(grid[nr][nc])) {
                    hasKey[grid[nr][nc] - 'a'] = true;
                    q.addAll(doors[grid[nr][nc] - 'a']);
                }

                q.offer(new int[]{nr, nc});
                vis[nr][nc] = true;
            }
        }

        return cnt;
    }
}

File: 2026-01/src/day_28/boj_9328/Main.cpp
#include <bits/stdc++.h>
using namespace std;

int dr[4] = {-1, 0, 1, 0};
int dc[4] = {0, 1, 0, -1};
char grid[102][102];
bool vis[102][102];
bool hasKey[26];

int bfs(int n, int m) {
    queue<pair<int, int>> q;
    q.push({0, 0});

    vector<queue<pair<int, int>>> doors(26);

    vis[0][0] = true;

    int cnt = 0;

    while (!q.empty()) {
        auto [r, c] = q.front();
        q.pop();

        if (grid[r][c] == '$') cnt++;

        for (int d = 0; d < 4; d++) {
            int nr = r + dr[d];
            int nc = c + dc[d];

            if (nr < 0 || nr >= n || nc < 0 || nc >= m) continue;
            if (grid[nr][nc] == '*' || vis[nr][nc]) continue;

            if ('A' <= grid[nr][nc] && grid[nr][nc] <= 'Z') {
                if (!hasKey[grid[nr][nc] - 'A']) {
                    doors[grid[nr][nc] - 'A'].push({nr, nc});
                    continue;
                }
            }

            if ('a' <= grid[nr][nc] && grid[nr][nc] <= 'z') {
                hasKey[grid[nr][nc] - 'a'] = true;
                while (!doors[grid[nr][nc] - 'a'].empty()) {
                    q.push(doors[grid[nr][nc] - 'a'].front());
                    doors[grid[nr][nc] - 'a'].pop();
                }
            }

            q.push({nr, nc});
            vis[nr][nc] = true;
        }
    }

    return cnt;
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int t;
    cin >> t;

    while (t--) {
        int h, w;
        cin >> h >> w;

        int n = h + 2;
        int m = w + 2;

        memset(vis, 0, sizeof(vis));
        for (int i = 0; i < n; i++) {
            grid[i][0] = grid[i][m - 1] = '.';
        }
        for (int j = 0; j < m; j++) {
            grid[0][j] = grid[n - 1][j] = '.';
        }

        for (int i = 1; i <= h; i++) {
            string s;
            cin >> s;

            for (int j = 1; j <= w; j++) {
                grid[i][j] = s[j - 1];
            }
        }

        string s;
        cin >> s;

        memset(hasKey, 0, sizeof(hasKey));
        if (!(s == "0")) {
            for (char c : s) {
                hasKey[c - 'a'] = true;
            }
        }

        cout << bfs(n, m) << '\n';
    }
}
